### PR TITLE
Review for alternative to #169

### PR DESCRIPTION
**Pull request by ted-miller**
I'm just creating this PR as a way to review an alternate fork/branch to #169. 

This is in response to: https://github.com/Yaskawa-Global/motoros2/pull/169#discussion_r1384674728

### CODE REVIEW

File: src/ControllerStatusIO.c [hunk 1]
```diff
@@ -114,23 +120,28 @@ BOOL Ros_Controller_Initialize()
     }
 
     //get the robot calibration data for multi-robot systems
-    for (i = 0; i < MAX_ROBOT_CALIBRATION_FILES; i += 1)
+    const BOOL bCalibLoadedOk = Ros_Controller_LoadGroupCalibrationData(&g_Ros_Controller);
+    //see whether the user should be notified about failures (it's OK to not
+    //have/load any calibration in some cases)
+    const BOOL bNeedToWarnAboutCalib = Ros_Controller_ShouldWarnNoCalibDataLoaded(
+        &g_Ros_Controller, g_nodeConfigSettings.publish_tf, bCalibLoadedOk);
```
Comment (ted-miller) [inline src/ControllerStatusIO.c:127]: The last two arguments are swapped. 

```suggestion
        &g_Ros_Controller, bCalibLoadedOk, g_nodeConfigSettings.publish_tf);
```

File: src/ControllerStatusIO.c [hunk 2]
```diff
@@ -114,23 +120,28 @@ BOOL Ros_Controller_Initialize()
...
-                MP_COORD* coord = &g_Ros_Controller.ctrlGroups[groupIndex]->robotCalibrationToBaseFrame;
-                coord->x = calibData.pos_uow[0];
-                coord->y = calibData.pos_uow[1];
-                coord->z = calibData.pos_uow[2];
-                coord->rx = calibData.ang_uow[0];
-                coord->ry = calibData.ang_uow[1];
-                coord->rz = calibData.ang_uow[2];
-            }
+            Ros_Debug_BroadcastMsg(
+                "%s: ignoring absence of calibration data (or load failure) as configured");
```
Comment (ted-miller) [inline src/ControllerStatusIO.c:136]: `%s` exists, but there's no argument

File: src/ControllerStatusIO.c [hunk 3]
```diff
@@ -114,23 +120,28 @@ BOOL Ros_Controller_Initialize()
...
+        }
+        else
+        {
+            mpSetAlarm(ALARM_CONFIGURATION_FAIL, "No calibration: check TF",
```
Comment (ted-miller) [inline src/ControllerStatusIO.c:140]:
```suggestion
            mpSetAlarm(ALARM_CONFIGURATION_FAIL, "No calibration: invalid /tf",
```

File: src/ControllerStatusIO.c [hunk 4]
```diff
@@ -114,23 +120,28 @@ BOOL Ros_Controller_Initialize()
...
+                SUBCODE_CONFIGURATION_NO_CALIB_FILES_LOADED);
+            Ros_Debug_BroadcastMsg(
+                "%s: no calibration files loaded: TF potentially incorrect -> warn "
```
Comment (ted-miller) [inline src/ControllerStatusIO.c:143]: To me, the `-> warn` is confusing. I'd suggest just doing something like `warning: no calibration...`

File: src/ControllerStatusIO.c [hunk 5]
```diff
@@ -751,3 +762,86 @@ int Ros_Controller_GetActiveAlarmCodes(USHORT active_alarms[MAX_ALARM_COUNT + MA
...
+                coord->ry = calibData.ang_uow[1];
+                coord->rz = calibData.ang_uow[2];
+            }
+        }
+    }
+
+    return bOneCalibFileLoaded;
+}
+
+//TODO(gavanderhoorn): should be able to make this 'static' once Yaskawa-Global/motoros2#173
```
Comment (ted-miller) [inline src/ControllerStatusIO.c:815]: I don't understand this comment. Why can't it be static now?

File: src/ControllerStatusIO.h [hunk 6]
```diff
@@ -142,6 +142,8 @@ extern int Ros_Controller_GetAlarmCode();
 //retrieve all active alarms (and a possible error) and store them in 'active_alarms'
 extern int Ros_Controller_GetActiveAlarmCodes(USHORT active_alarms[MAX_ALARM_COUNT + MAX_ERROR_COUNT]);
 
+//TODO(gavanderhoorn): make static, see comment on definition
+extern BOOL Ros_Controller_ShouldWarnNoCalibDataLoaded(Controller const* controller, BOOL bCalibLoadedOk, BOOL bPublishTfEnabled);
```
Comment (ted-miller) [inline src/ControllerStatusIO.h:146]: I don't get why this is in the 'public' header

File: src/MotoROS.h [hunk 7]
```diff
@@ -101,6 +101,7 @@
 #include "Tests_CtrlGroup.h"
 #include "Tests_TestUtils.h"
 #include "Tests_RosMotoPlusConversionUtils.h"
+#include "Tests_ControllerStatusIO.h"
```
Comment (ted-miller) [inline src/MotoROS.h:104]: The source and header for this has not been added to the VS project.

File: src/ControllerStatusIO.c (hunk 1 above)
Comment (gavanderhoorn) [reply src/ControllerStatusIO.c:127]: Good catch, thanks.

File: src/ControllerStatusIO.c (hunk 4 above)
Comment (gavanderhoorn) [reply src/ControllerStatusIO.c:143]: I'll rephrase it, but this is logging whether a warning should be displayed. The `->` is a logic implication.

File: src/ControllerStatusIO.c (hunk 5 above)
Comment (gavanderhoorn) [reply src/ControllerStatusIO.c:815]: `static` functions are local to the compilation unit. The test calling this function is not in this compilation unit. If it were `static`, we could not call it from elsewhere.

File: src/ControllerStatusIO.h (hunk 6 above)
Comment (gavanderhoorn) [reply src/ControllerStatusIO.h:146]: See my earlier comment.

File: src/MotoROS.h (hunk 7 above)
Comment (gavanderhoorn) [reply src/MotoROS.h:104]: That was on purpose, as these tests still hang my YRC.

**Comment by gavanderhoorn** [comment]
Thanks for the review @ted-miller.

I'll update the original #169 and close this one. We can continue there.

File: src/MotoROS.h (hunk 7 above)
Comment (gavanderhoorn) [reply src/MotoROS.h:104]: I guess you posted this as it breaks the MSBuild build.

That's indeed problematic. Will fix/work-around.

**Pull request closed without merging**